plutus-core/plutus-core/src/PlutusCore/Default/Universe.hs

```diff
@@ -339,90 +336,142 @@ instance TestTypesFromTheUniverseAreAllKnown DefaultUni
-            -- OPTIMIZE: benchmark an alternative `integerToIntMaybe`, modified from 'ghc-bignum'
-            if fromIntegral (minBound :: Int64) <= i && i <= fromIntegral (maxBound :: Int64)
-                then pure $ fromIntegral i
-                else throwing_ _EvaluationFailure
+            -- TODO: benchmark an alternative 'integerToIntMaybe', modified from @ghc-bignum@
+            if fromIntegral (minBound :: a) <= i && i <= fromIntegral (maxBound :: a)
+                then pure . AsInteger $ fromIntegral i
+                else throwing _OperationalUnliftingError . MkUnliftingError $ fold
+                        [ Text.pack $ show i
+                        , " is not within the bounds of "
+                        , Text.pack . show . typeRep $ Proxy @a
+                        ]
     {-# INLINE readKnown #-}
 
 #if WORD_SIZE_IN_BITS == 64
 -- See Note [Integral types as Integer].
-
-instance KnownTypeAst tyname DefaultUni Int where
-    toTypeAst _ = toTypeAst $ Proxy @Integer
-
-instance HasConstantIn DefaultUni term => MakeKnownIn DefaultUni term Int where
-    -- Convert Int-to-Integer via Int64.  We could go directly `toInteger`, but this way
-    -- is more explicit and it'll turn into the same thing anyway.
-    -- Although this conversion is safe regardless of the CPU arch (unlike the opposite conversion),
-    -- we constrain it to 64-bit for the sake of uniformity.
-    makeKnown = makeKnown . fromIntegral @Int @Int64
-    {-# INLINE makeKnown #-}
-
+deriving via AsInteger Int instance
+    KnownTypeAst tyname DefaultUni Int
+deriving via AsInteger Int instance HasConstantIn DefaultUni term =>
+    MakeKnownIn DefaultUni term Int
 instance HasConstantIn DefaultUni term => ReadKnownIn DefaultUni term Int where
-    -- Convert Integer-to-Int via Int64. This instance is safe only for 64-bit architecture
-    -- where Int===Int64 (i.e. no truncation happening).
     readKnown term = fromIntegral @Int64 @Int <$> readKnown term
     {-# INLINE readKnown #-}
-#endif
-
-instance KnownTypeAst tyname DefaultUni Word8 where
-    toTypeAst _ = toTypeAst $ Proxy @Integer
-
--- See Note [Integral types as Integer].
-instance HasConstantIn DefaultUni term => MakeKnownIn DefaultUni term Word8 where
-    makeKnown = makeKnown . toInteger
-    {-# INLINE makeKnown #-}
 
-instance HasConstantIn DefaultUni term => ReadKnownIn DefaultUni term Word8 where
-    readKnown term =
-        inline readKnownConstant term >>= oneShot \(i :: Integer) ->
-           case toIntegralSized i of
-               Just w8 -> pure w8
-               _       -> throwing_ _EvaluationFailure
+deriving via AsInteger Word instance
+    KnownTypeAst tyname DefaultUni Word
+deriving via AsInteger Word instance HasConstantIn DefaultUni term =>
+    MakeKnownIn DefaultUni term Word
+instance HasConstantIn DefaultUni term => ReadKnownIn DefaultUni term Word where
+    readKnown term = fromIntegral @Word64 @Word <$> readKnown term
```

Comment:
This I think should be inside the `#if WORD_SIZE_IN_BITS == 64  cpp macro` since you are assuming Word is 64bits?

Reply:
It is, it's just that the diff in unwieldy.

![Screenshot from 2024-05-30 13-22-32](https://github.com/IntersectMBO/plutus/assets/10480926/6b6902b9-5d80-4afc-be4a-09fa9d8af0a2)

Reply:
oups sry